container_opts.go

```diff
@@ -255,9 +223,19 @@ func WithSnapshotCleanup(ctx context.Context, client *Client, c containers.Conta
 	return nil
 }
 
+// WithNewSnapshot allocates a new snapshot to be used by the container as the
+// root filesystem in read-write mode
+func WithNewSnapshot(id string, i Image, opts ...snapshots.Opt) NewContainerOpts {
+	return withNewSnapshotBase(id, i, false, opts...)
+}
+
 // WithNewSnapshotView allocates a new snapshot to be used by the container as the
 // root filesystem in read-only mode
 func WithNewSnapshotView(id string, i Image, opts ...snapshots.Opt) NewContainerOpts {
+	return withNewSnapshotBase(id, i, true, opts...)
+}
+
+func withNewSnapshotBase(id string, i Image, readonly bool, opts ...snapshots.Opt) NewContainerOpts {
 	return func(ctx context.Context, client *Client, c *containers.Container) error {
 		diffIDs, err := i.(*image).i.RootFS(ctx, client.ContentStore(), client.platform)
```

Comment:
For `WithNewSnapshot`, the behavior seems to have changed here, but *i haven't verified that*.
https://github.com/containerd/containerd/blob/e7276fe35a6f09fdb7cacfdcfc5049b830f8da1c/cmd/ctr/commands/run/run_unix.go#L150-L156

The client and image's platform may not be the same.

Reply:
great call out. I think you're right, and we should follow more the `WithNewSnapshot`'s impl instead of `WithNewSnapshotView`'s (i.e., using `i.RootFS(ctx)` instead of `i.(*image).i.RootFS(ctx, client.ContentStore(), client.platform)` to get diffids), since `The client and image's platform may not be the same.` also applies to the `WithNewSnapshotView` func.

Reply:
🤔️Similarly, maybe we should avoid all uses of [`i.(*image).i.RootFS(ctx, client.ContentStore(), client.platform)`](https://github.com/search?q=repo%3Acontainerd%2Fcontainerd+%22i.%28*image%29.i.RootFS%22&type=code) since they all assume `client and image have the same platform`, which may not true if the image is created via `NewImageWithPlatform` as in your example.